pkg/sql/catalog/randgen/tables.go

```diff
@@ -0,0 +1,197 @@
+		namesPerTemplate++
+	}
+
+	// Names/templates, per generator.
+	startTmpl := make([]*tbTemplate, 0, num)
+	startNames := make([][]string, 0, num)
+	if g.tbNamePat == "_" {
+		// Rotate through templates and their names.
+		for i := range g.templates {
+			tbNamePat := g.templates[i].namePat
+			cfg := g.cfg.NameGen
+			if namesPerTemplate == 1 {
+				cfg.Number = false
+				if !cfg.HasVariability() {
+					cfg.Number = true
+				}
+			}
+			ng := randident.NewNameGenerator(&cfg, g.rand, tbNamePat)
+			tbNames, err := ng.GenerateMultiple(ctx, namesPerTemplate, conflictNames)
+			if err != nil {
+				panic(genError{err})
+			}
+			startNames = append(startNames, tbNames)
+			startTmpl = append(startTmpl, &g.templates[i])
+		}
+		// Now select the name,template pairs.
+		for nameOffset := 0; len(selectNames) < num; nameOffset++ {
+			for i := range startTmpl {
+				if len(selectNames) >= num {
+					break
+				}
+				selectTmpl = append(selectTmpl, startTmpl[i])
+				selectNames = append(selectNames, startNames[i][nameOffset])
+			}
+		}
+	} else {
+		// Rotate through templates, but use a single sequence of names.
+		ng := randident.NewNameGenerator(&g.cfg.NameGen, g.rand, g.tbNamePat)
+		tbNames, err := ng.GenerateMultiple(ctx, num, conflictNames)
+		if err != nil {
+			panic(genError{err})
+		}
+		selectNames = tbNames
+		for len(selectTmpl) < num {
+			for i := range g.templates {
+				if len(selectTmpl) >= num {
+					break
+				}
+				selectTmpl = append(selectTmpl, &g.templates[i])
+			}
+		}
+	}
+	return selectTmpl, selectNames
+}
+
+func (g *testSchemaGenerator) genOneTable(
+	ctx context.Context,
+	parentDb catalog.DatabaseDescriptor,
+	parentSc catalog.SchemaDescriptor,
+	privs *catpb.PrivilegeDescriptor,
+	id catid.DescID,
+	tmpl *tbTemplate,
+	tbName string,
+) {
+	defer func() {
+		if r := recover(); r != nil {
+			if e, ok := r.(genError); ok {
+				e.err = errors.Wrapf(e.err,
+					"generating table %s.%s.%s",
+					tree.NameString(parentDb.GetName()),
+					tree.NameString(parentSc.GetName()),
+					tree.NameString(tbName))
+				panic(e)
+			}
+			panic(r)
+		}
+	}()
+
+	tmpl.desc.ID = id
+	tmpl.desc.Name = tbName
+	tmpl.desc.ParentID = parentDb.GetID()
+	tmpl.desc.UnexposedParentSchemaID = parentSc.GetID()
+	tmpl.desc.Privileges = privs
+	tmpl.desc.Temporary = parentSc.SchemaKind() == catalog.SchemaTemporary
+	tmpl.desc.PrimaryIndex.Name = "primary"
+	if g.cfg.RandomizeColumns {
+		nameGenCfg := g.cfg.NameGen
+		nameGenCfg.Number = false
+		for i, cPat := range tmpl.baseColumnNames {
+			ng := randident.NewNameGenerator(&nameGenCfg, g.rand, cPat)
+			colName := ng.GenerateOne(0)
+			tmpl.desc.Columns[i+1].Name = colName
+			tmpl.desc.Families[0].ColumnNames[i+1] = colName
+		}
+		ng := randident.NewNameGenerator(&nameGenCfg, g.rand, "primary")
+		idxName := ng.GenerateOne(0)
+		tmpl.desc.PrimaryIndex.Name = idxName
+	}
+
+	tb := tabledesc.NewBuilder(&tmpl.desc).BuildCreatedMutableTable()
```

Comment:
You should call `RunPostDeserializationChanges` on the builder before building the descriptor here and elsewhere to make sure this functionality ages gracefully.